web/resource/resource.go

```diff
@@ -0,0 +1,178 @@
+// Package resource implements a RESTful resource handler
+package resource
+
+import (
+	"net/http"
+	"strings"
+
+	"darvaza.org/core"
+	"darvaza.org/x/web"
+	"darvaza.org/x/web/qlist"
+)
+
+var (
+	_ http.Handler = (*Resource[any])(nil)
+	_ web.Handler  = (*Resource[any])(nil)
+)
+
+// Resource is an http.Handler built around a given object
+// and a data type.
+type Resource[T any] struct {
+	check    CheckerFunc[T]
+	methods  []string
+	identity string
+
+	h  map[string]HandlerFunc[T]
+	r  map[string]HandlerFunc[T]
+	ql []qlist.QualityValue
+}
+
+// ServeHTTP handles the request initially using the TryServeHTTP method, and
+// then calling [web.HandleError] if there is any problem.
+func (r *Resource[T]) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if err := r.TryServeHTTP(rw, req); err != nil {
+		web.HandleError(rw, req, err)
+	}
+}
+
+// TryServeHTTP attempts to handle the request, and returns an error in the case of
+// problems instead of handling it locally.
+// This method converts panics into errors in case they occur.
+func (r *Resource[T]) TryServeHTTP(rw http.ResponseWriter, req *http.Request) error {
+	var c core.Catcher
+
+	return c.Do(func() error {
+		return r.doServeHTTP(rw, req)
+	})
+}
+
+func (r *Resource[T]) doServeHTTP(rw http.ResponseWriter, req *http.Request) error {
+	req, data, err := r.checkRequest(req)
+	if err != nil {
+		return err
+	}
+
+	h := r.getHandler(req)
+	if h == nil {
+		h = r.serve405
+	}
+
+	return h(rw, req, data)
+}
+
+func (r *Resource[T]) checkRequest(req *http.Request) (*http.Request, *T, error) {
+	req2, data, err := r.check(req)
+	switch {
+	case err != nil:
+		err = web.AsErrorWithCode(err, http.StatusBadRequest)
+		return nil, nil, err
+	case req2 != nil:
+		return req2, data, nil
+	default:
+		return req, data, nil
+	}
+}
+
+func (r *Resource[T]) getHandler(req *http.Request) HandlerFunc[T] {
+	method := strings.ToUpper(req.Method)
+	h := r.h[method]
+	return h
+}
+
+// Methods returns a list of all supported HTTP Methods
+func (r *Resource[T]) Methods() []string {
+	l := len(r.methods)
+	if l == 0 {
+		// early call.
+		return sortedKeys(r.h)
+	}
+
+	// return copy
+	out := make([]string, l)
+	copy(out, r.methods)
+	return out
+}
+
+func (r *Resource[T]) serve405(http.ResponseWriter, *http.Request, *T) error {
```

Comment:
See above

Reply:
405 is a request to a valid resource with an unsupported verb